source/iterator/iterator.go

```diff
@@ -0,0 +1,303 @@
+
+	// check if table exist.
+	rows, err := tx.QueryContext(ctx, fmt.Sprintf(queryIfExistTable, c.trackingTable))
+	if err != nil {
+		return fmt.Errorf("query exist table: %w", err)
+	}
+
+	defer rows.Close() //nolint:staticcheck,nolintlint
+
+	for rows.Next() {
+		var count int
+		er := rows.Scan(&count)
+		if er != nil {
+			return fmt.Errorf("scan: %w", err)
+		}
+
+		if count == 1 {
+			// table exist, setup not needed.
+			return nil
+		}
+	}
+
+	// create tracking table with all columns from `table`
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(queryCreateTable, c.trackingTable, c.table))
+	if err != nil {
+		return fmt.Errorf("create tracking table: %w", err)
+	}
+
+	// add columns to tracking table.
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(queryAddColumns, c.trackingTable, columnOperationType,
+		columnTimeCreated, columnTrackingID))
+	if err != nil {
+		return fmt.Errorf("add columns: %w", err)
+	}
+
+	// set not null for tracking id column.
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(querySetNotNull, c.trackingTable, columnTrackingID))
+	if err != nil {
+		return fmt.Errorf("set not null: %w", err)
+	}
+
+	// generate identity for tracking id column.
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(querySetGeneratedIdentity, c.trackingTable, columnTrackingID))
+	if err != nil {
+		return fmt.Errorf("generate identity: %w", err)
+	}
+
+	// reorg table.
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(queryReorgTable, c.trackingTable))
+	if err != nil {
+		return fmt.Errorf("reorganize table: %w", err)
+	}
+
+	// add index.
+	_, err = tx.ExecContext(ctx, fmt.Sprintf(queryAddIndex, c.table, c.trackingTable, columnTrackingID))
+	if err != nil {
+		return fmt.Errorf("add index: %w", err)
+	}
+
+	triggersQuery := buildTriggers(c.trackingTable, c.table, c.columnTypes)
+
+	// add trigger to catch insert.
+	_, err = tx.ExecContext(ctx, triggersQuery.queryTriggerCatchInsert)
+	if err != nil {
+		return fmt.Errorf("add trigger catch insert: %w", err)
+	}
+
+	// add trigger to catch update.
+	_, err = tx.ExecContext(ctx, triggersQuery.queryTriggerCatchUpdate)
+	if err != nil {
+		return fmt.Errorf("add trigger catch update: %w", err)
+	}
+
+	// add trigger to catch delete.
+	_, err = tx.ExecContext(ctx, triggersQuery.queryTriggerCatchDelete)
+	if err != nil {
+		return fmt.Errorf("add trigger catch delete: %w", err)
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return fmt.Errorf("commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+// If the underlying snapshot iterator returns false, the combined iterator will try to switch to the cdc iterator.
+func (c *CombinedIterator) HasNext(ctx context.Context) (bool, error) {
+	switch {
+	case c.snapshot != nil:
+		hasNext, err := c.snapshot.HasNext(ctx)
+		if err != nil {
+			return false, fmt.Errorf("snapshot has next: %w", err)
+		}
+
+		if !hasNext {
+			if er := c.switchToCDCIterator(ctx); er != nil {
```

Comment:
Why not use `err` instead of `er`?

Reply:
Because on this case we shadow variable err